File: poclb/iadt.c
/*
 * Copyright (c) 2020 NVI, Inc.
 *
 * This file is part of VLBI Field System
 * (see http://github.com/nvi-inc/fs).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
/* iadt

   Add an increment to the standard rte time array
   Copied from FORTRAN.
*/

static int itm[5]={100,60,60,24,366};
/* itm holds the maximum for each unit of time */

void iadt(it,idt,ires)
int it[6];     /* standard rte time array   */
int idt;       /* the increment to be added */
int ires;      /* resolution of idt, i.e. which unit of time applies */

{
  int i,iix;  

/* number of days in the year plus one */

  itm[4] = 366;
  /* not Y2.1K compliant */
  if (it[5]%4 == 0)
    itm[4]=367;

  iix = idt;

  for (i=ires-1; i<=4; i++) {
    it[i]=it[i]+iix;           /* increment */
    if (it[i] >= itm[i]) {     /* this unit has overflowed */
      if (i != 4) {            /* get the remainder for this unit */
         iix=it[i]/itm[i];        /* get the carry to the next higher unit */
         it[i]=it[i]%itm[i];      /* set current units to the remainder */ 
      } else
         it[4]=1+it[4]%itm[4];
    } else
      return;
  }
  it[5]=it[5]+1;  /* If we got here, all units have overflowed, up to the
                   * number of days in a year. Assume we haven't overflowed
                   * more than one year! */ 
}

File: fslb/iadt.f
*
* Copyright (c) 2020 NVI, Inc.
*
* This file is part of VLBI Field System
* (see http://github.com/nvi-inc/fs).
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*
      subroutine iadt(it,idt,ires)!c#870407:13:00    hp add time#

      dimension it(1),itm(5)
      data itm/100,60,60,24,366/
C
C   GET INITIAL INCREMENT
C
      iix=idt
C
C   INITIALIZE ILP
C   GET # DAYS/YEAR PLUS 1
C
      itm(5)=366
c not Y2.1K compliant
      if (mod(it(6),4).eq.0) itm(5)=367
C
C   SET UP LOOP
C
      do 20000 i=ires,5
C
C   INCREMENT THIS VALUE
C 
        it(i)=it(i)+iix 
C 
C   HAS THIS UNIT OVERFLOWED
C 
        if (it(i).lt.itm(i)) goto 60000
        if (i.ne.5) then
C 
C   GET THE CARRY TO HIGHER UNIT
C 
          iix=it(i)/itm(i)
C 
C   GET REMAINDER FOR THIS UNIT 
C 
          it(i)=mod(it(i),itm(i)) 
        else
          it(5)=1+mod(it(5),itm(5)) 
        endif
20000   continue
C 
C   INCREMENT YEAR (ASSUME NO BIGGER CHANGE)
C 
      it(6)=it(6)+1 
C 
60000 continue
C 
      return
      end 
